### DISCUSSION TITLE
发送请求都是400 Bad Request returned

### DISCUSSION THREAD

**Discussion by xinlinnet** [w7corp/easywechat#2410] [General]
HTTP/1.1 400 Bad Request returned for "https://api.mch.weixin.qq.com/v3/pay/transactions/jsapi" 是哪儿请求不正确吗，其他接口也 一样

**Comment by overtrue** [comment]
请提供上下文及安装的版本

**Reply by xinlinnet** [reply]
easywechat 版本6.2.2。
$config = [
            'mch_id' => "",
            "appid" => "",
            // 商户证书
            'private_key' => realpath('certs/apiclient_key.pem'),
            'certificate' => realpath('certs/apiclient_cert.pem'),
            // v3 API 秘钥
            'secret_key' => '',
            // v2 API 秘钥
            'v2_secret_key' => '',
            'platform_certs' => [],
           'http' => [
               'base_uri' => 'https://api.mch.weixin.qq.com/',
            ],
        ];    
        
        $app = new Application($config);
        
        $data = [
            "mchid" => (string)$app->getConfig()['mch_id'],
            "out_trade_no" => 'xxxxxxxxxx',
            "appid" => "xxxxxxxxxxxxx",
            "description" => "Image形象店-深圳腾大-QQ公仔",
            "notify_url" => "xxxxxxxxxxxxxxxxx",
            "amount" => [
                "total" => 100,
                "currency" => "CNY"
            ],
            "payer" => [
                "openid" => ""
            ]
        ];
        
        $api = $app->getClient();
        $response = $api->post("v3/pay/transactions/jsapi", ['body'=>$data]);

**Reply by overtrue** [reply]
postJson 试试

**Reply by xinlinnet** [reply]
$response = $api->post("v3/pay/transactions/jsapi", ['body'=>$data]);

$response = $api->post("v3/pay/transactions/jsapi", ['json'=>$data]);

$response = $api->post("v3/pay/transactions/jsapi", ['json'=>json_encode($data)]);

都试过了。一样的

**Comment by jiangslee** [comment]
try catch看看exception报了啥错

**Reply by jiangslee** [reply]
@xinlinnet 
```
use Symfony\Component\HttpClient\Exception\ClientException;
use Symfony\Component\HttpClient\Exception\ServerException;

try{
    $data=[
                //省略...
    ];
    $res = $app->getClient->post(
        'v3/pay/transactions/native',
        ['json' => $data]
    )->toArray();
} catch (ServerException $err) {
    return json_decode($err->getResponse()->getContent(false), true);
} catch (ClientException $err) {
    return json_decode($err->getResponse()->getContent(false), true);
} catch (\Throwable $err) {
    return $err->getMessage();
}
```

**Reply by xinlinnet** [reply]
HTTP/1.1 400 Bad Request returned for "https://api.mch.weixin.qq.com/v3/pay/transactions/jsapi".
try catch了就是这个请求错误。版本是6.2.2

**Reply by jiangslee** [reply]
你按我上面的try catch试一下贴报错吧，具体的参数错误在ClientException有返回
<img width="895" alt="image" src="https://user-images.githubusercontent.com/27892476/160316054-547a7f1c-3b52-4e7b-9dee-cb384b1daf8e.png">

**Reply by jiangslee** [reply]
> HTTP/1.1 400 Bad Request returned for "https://api.mch.weixin.qq.com/v3/pay/transactions/jsapi".
> try catch了就是这个请求错误。版本是6.2.2

建议你用官方源，update下最新版本，`6.3.0+`已经支持在调用请求时单独设置`throw(false)`，微信的错误提示可以直接返回给开发者。

```shell
composer config -g repo.packagist composer https://repo.packagist.org
composer update w7corp/easywechat
```

文档：
https://www.easywechat.com/6.x/client.html#处理响应

**Comment by senmeng** [comment]
HTTP/1.1 400 Bad Request returned for "https://api.mch.weixin.qq.com/v3/pay/transactions/app".
同样的问题，postJson也试用还是报这个

**Reply by jiangslee** [reply]
建议你用官方源，update下最新版本，`6.3.0+`已经支持在调用请求时单独设置`throw(false)`，微信的错误提示可以直接返回给开发者。

```shell
composer config -g repo.packagist composer https://repo.packagist.org
composer update w7corp/easywechat
```

文档：
https://www.easywechat.com/6.x/client.html#处理响应
